persistence-api/src/main/java/io/stargate/db/cdc/SchemaAwareCDCProducer.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.db.cdc;
+
+import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import org.apache.cassandra.stargate.db.MutationEvent;
+import org.apache.cassandra.stargate.schema.TableMetadata;
+
+/**
+ * Represents a schema-aware CDC async producer.
+ *
+ * <p>Implementors will get the following guarantees:
+ *
+ * <ul>
+ *   <li>For a given table, {@link #createTableSchemaAsync} will be invoked before calling {@link
+ *       #publish} with a partition update for that table (awaiting for the creation future to
+ *       complete).
+ *   <li>In the same way, when table metadata change is first detected, {@link
+ *       #createTableSchemaAsync} will be invoked before calling {@link #publish}
+ * </ul>
+ */
+public abstract class SchemaAwareCDCProducer implements CDCProducer {
+  private final ConcurrentHashMap<UUID, TableSchemaManager> tableSchemaManager =
+      new ConcurrentHashMap<>();
+
+  @Override
+  public abstract CompletableFuture<Void> init(Map<String, Object> options);
+
+  /** Asynchronously publishes the schema of the table. */
+  protected abstract CompletableFuture<Void> createTableSchemaAsync(TableMetadata table);
+
+  /**
+   * Asynchronously publishes the partition update.
+   *
+   * <p>It will be invoked after {@link SchemaAwareCDCProducer#createTableSchemaAsync} was invoked
+   * for a given table and creation future was completed.
+   */
+  protected abstract CompletableFuture<Void> send(MutationEvent mutation);
+
+  @Override
+  public abstract CompletableFuture<Void> close();
+
+  @Override
+  public CompletableFuture<Void> publish(MutationEvent mutation) {
+    TableSchemaManager schemaManager =
+        tableSchemaManager.computeIfAbsent(
+            mutation.getTable().getId(), k -> new TableSchemaManager(this::createTableSchemaAsync));
+
+    return schemaManager
+        .ensureCreated(mutation.getTable())
+        // Invoke send() with the version that later affected this table,
+        // not the most recent version number.
+        .thenCompose(originalSchemaVersion -> send(mutation));
```

Comment:
I guess this can lead to mutation events piling up in the `CompletableFuture` dependency graph if the schema creation is taking a long time and incoming traffic is heavy... should we perhaps limit this somehow? Should we perhaps skip mutations until the schema has been created?

Reply:
We shouldn't skip mutations because from the user pov, if cdc is enabled on a table it should either succeed or throw.

We can add mitigations in the future, some of them we already discussed w/ @tomekl007:
- Set creation timeouts.
- Reject after n queued mutations.
- Try to warmup the schema creation when a new stargate node is added (to minimize creations).

I think this mitigations are out of scope for the first iteration of CDC.

Reply:
sgtm